### ISSUE TITLE
Offline Map Properties not being saved on crash

### ISSUE THREAD

**Issue by jessicagood** [StraboSpot/strabo-mobile#493]
When downloading tiles for an offline map each tile is written to the local db after it is downloaded. 
However, the map properties, which among other things, states which downloaded tiles belong to that map, are not written to the local db until all the tiles are downloaded. Thus when the app crashes it may have downloaded (and saved to the local db) a number of tiles already but the map properties does not know this since the app crashed before the properties were written to the local db. I can write the map properties to the local db after every tile is downloaded but this doubles the number of db calls and is crazy when we get to downloading thousands of tiles. In this case maybe it's better to download tiles in chunks from the StraboSpot server as you suggested @jasonash?

I did already make a fix so the map name is written to the local db before download begins (see commit da69786b49ddcd6d6947f8b1387fa53cf165f961) so at least that's not an issue any longer.

**Comment by jasonash** [comment]
@jessicagood Yes, I think @panamint55 suggested trying to download in chunks which might alleviate the crashing issue. Maybe we could start with groups of 200 to see how that works?

**Comment by jessicagood** [comment]
If I do it from my end we're still making 200 server calls then 400 db calls (to save the tile and then map properties each time). Sometimes (rarely) @panamint55 said he could get the app to crash in less than 200 calls. 200 is pretty arbitrary so I'm not sure if that's a fix that will be consistent.

**Comment by jasonash** [comment]
@jessicagood So you're making a server call for every tile? I thought there was some way we could chunk them into groups of 200 for asynchronous calls. If we're making separate calls for every tile, maybe we should investigate other options.

**Comment by jessicagood** [comment]
We're eventually going to need to hit our own tile servers instead of osm and mapbox anyway as we get more users too since they have limits on how many tiles you can download. OSM doesn't even cache tiles at zoom 17 and above, those tiles are rendered on their servers each time a request is made. 
https://operations.osmfoundation.org/policies/tiles/
Mapbox's tile ceiling is 6000 tiles https://www.mapbox.com/help/mobile-offline/

**Comment by jasonash** [comment]
OK, I'll start going over the app code for tile download and see if I can port it to Python for server-side collection/caching. This will have the added benefit of being much faster since we'll only have to download once. Have you experimented at all with .zip archives within Angular? I'm hoping that isn't another roadblock.

**Comment by jessicagood** [comment]
With commit 72bf5b774734fecb60c134e7976fdc052bf154a0 I am saving the map properties after each tile is download so in case of a crash we still know which downloaded tiles belong to the map. This will make the download process a little slower but will give us a safeguard if there is a crash. We can still restart the app and append the tiles that didn't get downloaded the first time.

**Comment by jasonash** [comment]
Maybe this will work? https://stuk.github.io/jszip/

**Comment by jessicagood** [comment]
Was just looking at that one too. I haven't used zip files with Angular before but looks like it is doable. We can also see if we get better performance off of hitting our own tile servers before making the jump to zipping chunks.
